samples/snippets/deny_policies.py

```diff
@@ -0,0 +1,325 @@
+      Params:
+      project_id: ID or number of the Google Cloud project you want to use.
+      policy_id: Specify the id of the Deny policy you want to create.
+    """
+    policies_client = iam_v2beta.PoliciesClient()
+
+    # Each deny policy is attached to an organization, folder, or project.
+    # To work with deny policies, specify the attachment point.
+    #
+    # Its format can be one of the following:
+    # 1. cloudresourcemanager.googleapis.com/organizations/ORG_ID
+    # 2. cloudresourcemanager.googleapis.com/folders/FOLDER_ID
+    # 3. cloudresourcemanager.googleapis.com/projects/PROJECT_ID
+    #
+    # The attachment point is identified by its URL-encoded resource name. Hence, replace
+    # the "/" with "%2F".
+    attachment_point = f"cloudresourcemanager.googleapis.com%2Fprojects%2F{project_id}"
+
+    deny_rule = types.DenyRule()
+    # Add one or more principals who should be denied the permissions specified in this rule.
+    # For more information on allowed values, see: https://cloud.google.com/iam/docs/principal-identifiers
+    deny_rule.denied_principals = ["principalSet://goog/public:all"]
+
+    # Optionally, set the principals who should be exempted from the
+    # list of denied principals. For example, if you want to deny certain permissions
+    # to a group but exempt a few principals, then add those here.
+    # deny_rule.exception_principals = ["principalSet://goog/group/project-admins@example.com"]
+
+    # Set the permissions to deny.
+    # The permission value is of the format: service_fqdn/resource.action
+    # For the list of supported permissions, see: https://cloud.google.com/iam/help/deny/supported-permissions
+    deny_rule.denied_permissions = [
+        "cloudresourcemanager.googleapis.com/projects.delete"
+    ]
+
+    # Optionally, add the permissions to be exempted from this rule.
+    # Meaning, the deny rule will not be applicable to these permissions.
+    # deny_rule.exception_permissions = ["cloudresourcemanager.googleapis.com/projects.create"]
+
+    # Set the condition which will enforce the deny rule.
+    # If this condition is true, the deny rule will be applicable. Else, the rule will not be enforced.
+    # The expression uses Common Expression Language syntax (CEL).
+    # Here we block access based on tags.
+    #
+    # A tag is a key-value pair that can be attached to an organization, folder, or project. You can use deny policies to deny permissions based on tags without adding an IAM Condition to every role grant.
+    # For example, imagine that you tag all of your projects as dev, test, or prod. You want only members of project-admins@example.com to be able to perform operations on projects that are tagged prod.
+    # To solve this problem, you create a deny rule that denies the cloudresourcemanager.googleapis.com/projects.delete permission to everyone except project-admins@example.com for resources that are tagged test.
+    deny_rule.denial_condition = expr_pb2.Expr(
```

Comment:
I was surprised to see that you had to import the proto in order to do this; I would have guessed that there was a more Pythonic wrapper. That said, I'm also not very familiar with the Cloud Client Libraries for Python.

Is this pretty typical, or is it unusual that customers would need to use the protos directly?

Reply:
I've seen code samples which invoke proto fields this way. The class and functions come from "googleapis-common-protos" library.
Hence, I am guessing this is a common way!

Reply:
Have you tried instantiating deny_rule.denial_condition directly, without using the proto? This would be preferred; a lot of end users get confused by protos.

Reply:
This line: https://github.com/googleapis/python-iam/blob/main/google/cloud/iam_v2beta/types/deny.py#L102
states that the argument to the denial_condition is of type google.type.expr_pb2.Expr

I am not sure how else this can be initalized. Please lmk.